src/transcodeImage.js

```diff
@@ -0,0 +1,283 @@
+  return transcodeMap[transferSyntaxUid];
+}
+
+/**
+ * Transcode imageFrame. It uses transcodeMap to define the target encoding based on source encoding.
+ *
+ * @param {*} id object containing transferSyntaxUid
+ * @param {*} targetId target object containing transferSyntaxUid
+ * @param {*} imageFrame data to be transcoded
+ * @param {*} dataSet related image frame dataset
+ * @param {*} options runner options
+ * @returns object result for transcoding operation with id and image frame.
+ */
+async function transcodeImageFrame(
+  id,
+  targetId,
+  imageFrame,
+  dataSet,
+  options = {}
+) {
+  let result = {};
+
+  if (!shouldTranscodeImageFrame(id, options)) {
+    return {
+      id,
+      imageFrame: imageFrame,
+      done: false,
+    };
+  }
+
+  const transcoder = getTranscoder(id.transferSyntaxUid);
+
+  // last chance to prevent transcoding
+  if (targetId.transferSyntaxUid !== transcoder.transferSyntaxUid) {
+    return {
+      id,
+      imageFrame: imageFrame,
+      done: false,
+    };
+  }
+
+  const imageInfo = getImageInfo(dataSet);
+  let done = false;
+  let processResultMsg = "";
+
+  try {
+    switch (transcoder.transcodeOp) {
+      case transcodeOp.transcode:
+        transcodeLog(
+          options,
+          `Full transcoding image from \x1b[43m${id.transferSyntaxUid}\x1b[0m to \x1b[43m${targetId.transferSyntaxUid}\x1b[0m`
+        );
+
+        result = await dicomCodec.transcode(
+          imageFrame,
+          imageInfo,
+          id.transferSyntaxUid,
+          targetId.transferSyntaxUid
+        );
+
+        processResultMsg = `Transcoding finished`;
+        break;
+      case transcodeOp.encodeOnly:
+        transcodeLog(
+          options,
+          `Encoding image to \x1b[43m${targetId.transferSyntaxUid}\x1b[0m`
+        );
+
+        result = await dicomCodec.encode(
+          imageFrame,
+          imageInfo,
+          targetId.transferSyntaxUid
+        );
+
+        processResultMsg = `Encoding finished`;
+        break;
+      case transcodeOp.decodeOnly:
+        transcodeLog(
+          options,
+          `Decoding image from \x1b[43m${id.transferSyntaxUid}\x1b[0m`
+        );
+        result = await dicomCodec.decode(
+          imageFrame,
+          imageInfo,
+          id.transferSyntaxUid
+        );
+
+        processResultMsg = `Decoding finished`;
+        break;
+    }
+
+    done = !!result.imageFrame;
+  } catch (e) {
+    transcodeLog(options, "Failed to transcode image", e);
+  }
+
+  // recover transfer syntax
+  if (!done) {
+    targetId = id;
+  }
+
+  // log final message only if resultMsg is present
+  if (processResultMsg) {
+    processResultMsg += done ? " successfully" : " unsuccessfully";
+    transcodeLog(options, processResultMsg);
+  }
+
+  let _imageFrame = result.imageFrame ?? imageFrame;
+
+  return {
+    id: targetId,
+    imageFrame: _imageFrame,
+    done,
+  };
+}
+
+function transcodeLog(options, msg, error = "") {
```

Comment:
This is quite simple but very functional, if a more robust logger is developed this should be moved elsewhere